security/src/main/scala/pl/touk/nussknacker/ui/security/oauth2/OAuth2ServiceFactory.scala

```diff
@@ -25,7 +26,24 @@ trait OAuth2Service[+UserInfoData, +AuthorizationData <: OAuth2AuthorizationData
       authorizationCode: String,
       redirectUri: String
   ): Future[(AuthorizationData, UserInfoData)]
-  def checkAuthorizationAndObtainUserinfo(accessToken: String): Future[(UserInfoData, Option[Instant])]
+  def checkAuthorizationAndObtainUserinfo(accessToken: String): Future[(UserInfoData, Option[Instant])] = {
+    for {
+      accessTokenData <- introspectAccessToken(accessToken)
+      userInfo        <- obtainUserInfo(accessToken, accessTokenData)
+    } yield (userInfo, accessTokenData.expirationTime)
+  }
+  def introspectAccessToken(accessToken: String): Future[IntrospectedAccessTokenData]
+  def obtainUserInfo(accessToken: String, accessTokenData: IntrospectedAccessTokenData): Future[UserInfoData]
+}
+
+final case class IntrospectedAccessTokenData(
+    subject: Option[String],
+    expirationTime: Option[Instant],
+    roles: Set[String]
+)
+
+object IntrospectedAccessTokenData {
+  val empty: IntrospectedAccessTokenData = IntrospectedAccessTokenData(None, None, Set.empty)
```

Comment:
I don't know if I get why we need to do it like this instead using `Option[IntrospectedAccessTokenData]` or `sealed trait` hierarchy like this:

```scala
sealed trait IntrospectedAccessTokenData
object IntrospectedAccessTokenData {
   final case class Instance(subject ...) extends IntrospectedAccessTokenData
   case object Empty extends IntrospectedAccessTokenData
}
```
```

Reply:
This is because our hierarchy structure of service classes is prepared for dozens of methods of authentication. For example we can handle access token that is not a JWT token (which is allowed by OAuth2 specification) and during access token introspection we can extract only some of the data (e.g. only `expirationTime`). And even if the access token is JWT, according to the specification, every claim is optional so we can have `subject` but no `expiration`. You might ask if we already use all these possibilities. I can't say it for 100% per cent. I know that we at least use authentication with OIDC (keycloak, Auth0), with GH which has some custom username claims and no option to define role claims and some custom implementation in at least one deployment of our product.